MonoGame.Framework/Graphics/Image2D.cs

```diff
@@ -0,0 +1,187 @@
+        public Color[] Data
+        {
+            get { return _data; }
+        }
+
+        /// <summary>
+        /// Color at the specified coordinate
+        /// </summary>
+        /// <param name="x"></param>
+        /// <param name="y"></param>
+        /// <returns></returns>
+        public Color this[int x, int y]
+        {
+            get
+            {
+                return _data[y * Width + x];
+            }
+
+            set
+            {
+                _data[y * Width + x] = value;
+            }
+        }
+
+
+        /// <summary>
+        /// Constructs a <see cref="Image2D"/> of the specified size
+        /// </summary>
+        /// <param name="width"></param>
+        /// <param name="height"></param>
+        public Image2D(int width, int height)
+        {
+            if (width <= 0)
+            {
+                throw new ArgumentOutOfRangeException("width");
+            }
+
+            if (height <= 0)
+            {
+                throw new ArgumentOutOfRangeException("height");
+            }
+
+            Width = width;
+            Height = height;
+
+            _data = new Color[width * height];
+        }
+
+        /// <summary>
+        /// Constructs a <see cref="Image2D"/> of the specified size and containing specified data
+        /// </summary>
+        /// <param name="width"></param>
+        /// <param name="height"></param>
+        /// <param name="data"></param>
+        public Image2D(int width, int height, Color[] data)
+        {
+            if (width <= 0)
+            {
+                throw new ArgumentOutOfRangeException("width");
+            }
+
+            if (height <= 0)
+            {
+                throw new ArgumentOutOfRangeException("height");
+            }
+
+            if (data == null)
+            {
+                throw new ArgumentNullException("data");
+            }
+
+            var length = width * height;
+            if (data.Length != length)
+            {
+                throw new ArgumentException(string.Format("Inconsistent data length: expected={0}, provided={1}", length,
+                    data.Length));
+            }
+
+            Width = width;
+            Height = height;
+            _data = data;
+        }
+
+        /// <summary>
+        /// Creates Texture2D from this image
+        /// </summary>
+        /// <param name="graphicsDevice"></param>
+        /// <returns></returns>
+        public Texture2D CreateTexture2D(GraphicsDevice graphicsDevice)
+        {
+            var texture = new Texture2D(graphicsDevice, Width, Height, false, SurfaceFormat.Color);
+            texture.SetData(_data);
+            return texture;
+        }
+
+        /// <summary>
+        /// Loads an image in PNG, JPG, BMP, GIF or TGA format from the stream
+        /// </summary>
+        /// <param name="stream"></param>
+        /// <returns></returns>
+        public unsafe static Image2D FromStream(Stream stream)
```

Comment:
Wait a second, you are reading a stream, than copying a stream to a memory stream, than copying memory stream to a byte array, than copying byte array to a color array.

Reply:
Well, the comment in the method code clearly indicates why stream is being copied to a memory stream.
Some platforms(like Xamarin) FileStream lack seek functionality, which caused #6507.
So in order to fix once and for all, I decided to copy whole stream into array of bytes and load it from there.